lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java

```diff
@@ -467,9 +467,19 @@ public TopDocs searchAfter(ScoreDoc after, Query query, int numHits) throws IOEx
 
     final CollectorManager<TopScoreDocCollector, TopDocs> manager = new CollectorManager<TopScoreDocCollector, TopDocs>() {
 
+      private HitsThresholdChecker hitsThresholdChecker;
       @Override
       public TopScoreDocCollector newCollector() throws IOException {
-        return TopScoreDocCollector.create(cappedNumHits, after, TOTAL_HITS_THRESHOLD);
+
+        if (hitsThresholdChecker == null) {
+          if (executor == null || leafSlices.length <= 1) {
+            hitsThresholdChecker = HitsThresholdChecker.create(TOTAL_HITS_THRESHOLD);
+          } else {
+            hitsThresholdChecker = HitsThresholdChecker.createShared(TOTAL_HITS_THRESHOLD);
+          }
+        }
```

Comment:
can we create it eagerly instead of lazily?

Reply:
Meaning, in the constructor? We will have to create a named implementation of the `CollectorManager` for that

Reply:
Fixed this by using `final` modifier